src/parseTools.js

```diff
@@ -985,6 +988,21 @@ function asmFFICoercion(value, type) {
 
 function makeDynCall(sig, funcPtr) {
   assert(sig.indexOf('j') == -1);
+  if (funcPtr === undefined) {
+    printErr(`warning: ${currentlyParsedFilename}: Legacy use of {{{ makeDynCall("${sig}") }}}(funcPtr, arg1, arg2, ...). Starting from Emscripten 2.0.2 (Aug 31st 2020), syntax for makeDynCall has changed. New syntax is {{{ makeDynCall("${sig}", "funcPtr") }}}(arg1, arg2, ...). Please update to new syntax.`);
+    var ret = (sig[0] == 'v') ? 'return ' : '';
+    var args = [];
+    for(var i = 1; i < sig.length; ++i) {
+      args.push(`a${i}`);
+    }
+    args = args.join(', ');
+
+    if (USE_LEGACY_DYNCALLS) {
+      return `(function(cb, ${args}) { return getDynCaller("${sig}", cb)(${args}) })`;
+    } else {
+      return `(function(cb, ${args}) { return wasmTable.get(cb)(${args}) })`;
+    }
+  }
```

Comment:
I love the clear error message, but actually handling the old form feels excessive to me. I suggest erroring the build, to force people to update? It's an internal API, and one that isn't hard to update for (with the nice new error message here).

Reply:
I am surprised to see you say that this is an internal API.

If this was an internal API (only used in Emscripten system library_*.js files), then indeed this would be excessive and I would not have minded too much about this, but would just fix our own library JS code in Unity and move on. However, `{{{ makeDynCall(..) }}}` works and is used also in user-provided library_*.js files, which makes this a public API.

Also we should not have core syntax in library_*.js that is not supposed to be used by end users. That is because most of the library_*.js files that people author are monkeyed off of existing src/library_*.js files - that is what we write to the GitHub issue tracker so many times in the past when people ask how to achieve something ("take a look at an example in src/library_webgl.js" etc). If `{{{ makeDynCall(..) }}}` was an internal API, it should certainly have been blocked from functioning in user library_*.js files at all. But why should user library_*.js files be somehow tiered to a lesser feature set than our own core library_*.js files?

Historically, according to my understanding, there have been three ways to invoke a dynamic pointer from JS code:
1. dynCall_sig(): static dispatch, works from library_*.js files and EM_ASM, EM_JS, --pre-js-, --post-js and external <script>s.
2. {{{ makeDynCall('sig) }}}: static dispatch, works from library_*.js files only. The "safe" and preferred way to do static dispatch since it works across all build modes (EMULATED_FUNCTION_POINTERS, USE_LEGACY_DYNCALLS).
3. dynCall('sig'): dynamic dispatch, works from library_*.js files and EM_ASM, EM_JS, --pre-js-, --post-js and external <script>s. Carries a significant performance hit compared to static dispatch.

PR #12059 regressed all of the syntaxes 1., 2. and 3. above (#12733).

This PR fixes the regression (2).

I do not know how to fix (1) at the moment, that is blocking us to be able to update Unity to a newer compiler version.

The regression of (3) is that `dynCall` was removed from the build by default, requiring library_*.js authors to now explicitly add a `__deps` field to it for it to work, or add it to `DEFAULT_LIBRARY_FUNCS_TO_INCLUDE` explicitly, if one wishes to do `dynCall()` from EM_ASM, EM_JS, --pre-js-, --post-js or external <script>s.

This regression (3) is minor for Unity's use, since we can paper over that in the build system easily, so not sweating too much about it (but it is still a breaking change that was not mentioned in the ChangeLog for Emscripten 2.0.2).

Chatting with @sbc100 on #12733 I believe there was a misassumption that (3) would be the "one syntax to rule them all". Historically I moved us away from the dynamic dispatch dynCall in Jan 2017 because of its large impact observed on real-world performance (#4894). We should not be going back to that.

Unity Asset Store has hundreds of packages that advertise WebGL support, and we have thousands to tens of thousands of users that develop their own WebGL project; of all of those an unknown number of packages and game projects WebGL packages with library_*.js files to interop with Unity C#, C/C++ and JS. We cannot break any of them when updating, because it may not even be up to the project developer to be able to fix up the library_*.js issue, but they must receive an update from the Asset Store package author.

If we errored out here in this message, that would mean that possibly thousands of users would be blocked from developing their game project until plugin authors provided a fix. That is not possible for us.

Because of these reasons is why I am adding a very explicit warning about this that will prompt developers who see the warning to look at their own library_*.js files, and nudge their plugin vendors to fix up theirs.

Reply:
I see, thanks @juj, I didn't realize this was used in user code so much. I'd assumed that low-level handling indirect calls is extremely rare (like it is in the emscripten JS libraries). Optimally user code would use something that does the low-level handling for it, but that's not the situation.

Given that, sounds good to support the old syntax.